readme.md
# redis-speed

An utility for redis aim to improve the performance of redis, it includes `FlashCacheRedis` now.

[![build status][travis-image]][travis-url]
[![Test coverage][coveralls-image]][coveralls-url]
[![David deps][david-image]][david-url]
[![node version][node-image]][node-url]

[npm-url]: https://npmjs.org/package/redis-speed
[travis-image]: https://img.shields.io/travis/yunnysunny/redis-speed.svg?style=flat-square
[travis-url]: https://travis-ci.org/yunnysunny/redis-speed
[coveralls-image]: https://img.shields.io/coveralls/yunnysunny/redis-speed.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/yunnysunny/redis-speed?branch=master
[david-image]: https://img.shields.io/david/yunnysunny/redis-speed.svg?style=flat-square
[david-url]: https://david-dm.org/yunnysunny/redis-speed
[node-image]: https://img.shields.io/badge/node.js-%3E=_6-green.svg?style=flat-square
[node-url]: http://nodejs.org/download/

[![NPM](https://nodei.co/npm/redis-speed.png?downloads=true)](https://nodei.co/npm/redis-speed/) 

## FlashCacheRedis  

Save value in memory to reduce the frequency of reading redis.

### Usage

```javascript
const {FlashCacheRedis} = require('redis-speed');
const Redis = require('ioredis');
const redisClient = new Redis();//connect to the redis server of localhost:6379
const {cacheQueryAdapter} = new FlashCacheRedis({
    redisClient,//the redis client object
    interval:1000
});
const {expect} = require('chai');
const KEY = 'flash_cache_redis:basic';
const VALUE = {name:'sunny',id:1};

redisClient.set(KEY,JSON.stringify(VALUE),function(err) {//save session
    if (err) {
        return console.error(err);
    }
    //It will cache the value in FlashCacheRedis instance, when you call the function of `cacheQueryAdapter.get` next, it will read the value directly from memory.
    cacheQueryAdapter.get(KEY,function(err,obj) {
        if (err) {
            return console.error(err);
        }
        try {
            obj = JSON.parse(obj);
        } catch (e) {
            return console.error(e);
        }
        expect(obj).to.have.property('name').and.equal(VALUE.name);
    });
});
```

## RedisHelper
An utility class to do parallel job with redis. If your redis is run in cluster mode, it will not return the data properly with some functions, such as `pipeline` or `mutil`. `RedisHelper` is designed to resoved such issue.

### Usage

```javascript
const {RedisHelper} = require('redis-speed');
const key1 = 'key1';
const value1 = Math.random() + '';
const key2 = 'key2';
const value2 = Math.random() + '';
const tasks = [
    ['set',key1,value1],
    ['set',key2,value2]
];
redisHelper.doParallelJobs(tasks,function(none,[[err1],[err2]]) {

});
```

## redisBatchIncr
An utility class to send redis command in batch.

### Usage

```javascript
const {
    redisBatchIncr: {
        EVENT_ONE_LOOP_FINISHED,
        EVENT_SEND_ERROR,
        BatchHincr,
        BatchZincrby
    }
} = require('redis-speed');
const redisClient = new Redis();//connect to the redis server of localhost:6379

const LOOP_COUNT = 100;
const INTERVAL = 200;
const key = ('test:' + Math.random()).replace('.','');
const cmd = new BatchHincr({redisClient,loopInterval:INTERVAL,key});

cmd.on(EVENT_ONE_LOOP_FINISHED,function() {
    redisClient.hget(key,'name',function(err,reply) {
        if (err) {
            return console.error(err);
        }
        expect(Number(reply)).to.be.equal(LOOP_COUNT);
    });
});
cmd.on(EVENT_SEND_ERROR,function(err) {
    console.error(err);
});

for(var i=0;i<LOOP_COUNT;i++) {
    cmd.addData(1,'name');
}
```

## redisMultiBatchIncr

An utility class to send  redis command in batch with multiple keys.

### Usage
```javascript
const {
    redisMultiBatchIncr: {
        EVENT_ONE_LOOP_FINISHED,
        EVENT_SEND_ERROR,
        BatchMultiHincr,
        BatchMultiZincrby
    }
} = require('redis-speed');
const redisClient = new Redis();//connect to the redis server of localhost:6379

const LOOP_COUNT = 100;
const INTERVAL = 200;

const cmd = new BatchMultiHincr({redisClient,loopInterval:INTERVAL});
const key = ('test:' + Math.random()).replace('.','');
cmd.on(EVENT_ONE_LOOP_FINISHED,function() {
    redisClient.hget(key,'name',function(err,reply) {
        if (err) {
            return done(err);
        }
        expect(Number(reply)).to.be.equal(LOOP_COUNT);
        done();
    });
});
cmd.on(EVENT_SEND_ERROR,function(err) {
    done(err);
});

for(var i=0;i<LOOP_COUNT;i++) {
    cmd.addData(key,1,'name');
}
```

## API
[api](docs/api.md)

## License
[MIT](LICENSE)





changelog.md
# v1.2.0
## Add
1. Add the function `expire` of `RedisHelper`.

# v1.1.0
## Add
1. Add the utility of `BatchMultiHincr`.

# v1.0.0
## Breaking changes
1. Remove the utility of `redisBatchCmd`.
2. You should give the key to the utility of `redisBatchIncr` when initial it.

# v0.4.0
## Add
1. Add the utility of `redisBatchIncr`, deprecate the `redisBatchCmd`.

# v0.3.0
## Add
1. Add some utility classes to send redis commands in batch way.

# v0.2.0
## Add 
1. Add class RedisHelper.

# v0.1.0
## Add
1. Project init.

test/mocha/basic_test.js
const Redis = require('ioredis');
const {expect} = require('chai');
const sinon  = require('sinon');
var mocha = require('mocha');
var describe = mocha.describe;
var it = mocha.it;
const {FlashCacheRedis} = require('../../index');
const redisClient = new Redis();//connect to the redis server of localhost:6379
const INTERVAL = 1000;
const {cacheQueryAdapter} = new FlashCacheRedis({
    redisClient,//the redis client object
    interval:INTERVAL,
    useYoungOnly:true
});
const KEY = 'flash_cache_redis:basic';
const VALUE = {name:'sunny',id:1};

describe('basic test',function() {
    it ('should set redis success',function(done) {
        redisClient.set(KEY,JSON.stringify(VALUE),function(err) {//save session
            if (err) {
                return done(err);
            }
            done();
        });
    });

    it ('should get success',function(done) {
        const redisGet = sinon.spy(redisClient,'get');
        cacheQueryAdapter.get(KEY,function(err,obj) {
            expect(redisGet.called).to.be.true;
            redisGet.restore();
            if (err) {
                return done(err);
            }
            try {
                obj = JSON.parse(obj);
            } catch (e) {
                return done(e);
            }
            expect(obj).to.have.property('name').and.equal(VALUE.name);
            done();
        });
    });

    it ('should get from memory success',function(done) {
        const redisGet = sinon.spy(redisClient,'get');
        cacheQueryAdapter.get(KEY,function(err,obj) {
            expect(redisGet.called).to.be.false;
            redisGet.restore();
            if (err) {
                return done(err);
            }
            try {
                obj = JSON.parse(obj);
            } catch (e) {
                return done(e);
            }
            expect(obj).to.have.property('name').and.equal(VALUE.name);
            done();
        });
    });

    it('delay',function(done) {
        setTimeout(function() {done();},INTERVAL+100);
    });

    it ('should get from redis success',function(done) {
        const redisGet = sinon.spy(redisClient,'get');
        cacheQueryAdapter.get(KEY,function(err,obj) {
            expect(redisGet.called).to.be.true;
            redisGet.restore();
            if (err) {
                return done(err);
            }
            try {
                obj = JSON.parse(obj);
            } catch (e) {
                return done(e);
            }
            expect(obj).to.have.property('name').and.equal(VALUE.name);
            done();
        });
    });
});

test/mocha/helper_test.js
const Redis = require('ioredis');
const {expect} = require('chai');
const async  = require('neo-async');
var mocha = require('mocha');
var describe = mocha.describe;
var it = mocha.it;
const {RedisHelper} = require('../../index');
const redisClient = new Redis();//connect to the redis server of localhost:6379
const EXPIRE_HOUR = 3;
const redisHelper = new RedisHelper({redisClient,expiredHour: EXPIRE_HOUR,expiredCacheTime:600 * 1000});

describe('helper test#', function() {
    it('do parallel in array',function(done) {
        const key1 = 'key1';
        const value1 = Math.random() + '';
        const key2 = 'key2';
        const value2 = Math.random() + '';
        async.waterfall([
            function(next) {
                const tasks = [
                    ['set',key1,value1],
                    ['set',key2,value2]
                ];
                redisHelper.doParallelJobs(tasks,function(none,[[err1],[err2]]) {
                    expect(err1).to.be.null;
                    expect(err2).to.be.null;
                    next();
                });
            },
            function(next) {
                const tasks = [
                    ['get',key1],
                    ['get',key2]
                ];
                redisHelper.doParallelJobs(tasks,function(none,[[err1,reply1],[err2,reply2]]) {
                    expect(err1).to.be.null;
                    expect(err2).to.be.null;
                    expect(reply1).to.be.equal(value1);
                    expect(reply2).to.be.equal(value2);
                    next();
                });
            }
        ],done);
        
    });

    it('do parallel in object',function(done) {
        const key1 = 'key1';
        const value1 = Math.random() + '';
        const key2 = 'key2';
        const value2 = Math.random() + '';
        async.waterfall([
            function(next) {
                const tasks = {
                    set1: ['set',key1,value1],
                    set2: ['set',key2,value2]
                };
                redisHelper.doParallelJobs(tasks,function(none,{
                    set1:[err1],
                    set2:[err2]
                }) {
                    expect(err1).to.be.null;
                    expect(err2).to.be.null;
                    next();
                });
            },
            function(next) {
                const tasks = {
                    get1:['get',key1],
                    get2:['get',key2]
                };
                redisHelper.doParallelJobs(tasks,function(none,{
                    get1:[err1,reply1],
                    get2:[err2,reply2]
                }) {
                    expect(err1).to.be.null;
                    expect(err2).to.be.null;
                    expect(reply1).to.be.equal(value1);
                    expect(reply2).to.be.equal(value2);
                    next();
                });
            }
        ],done);
        
    });

    it('do parallel with empty',function(done) {
        const key1 = 'key1';
        const value1 = Math.random() + '';
        const key2 = 'key2';
        const value2 = Math.random() + '';
        async.waterfall([
            function(next) {
                const tasks = [
                    ['set',key1,value1],
                    [RedisHelper.EMPTY_FUN],
                    ['set',key2,value2],
                ];
                redisHelper.doParallelJobs(tasks,function(none,[[err1],[errn],[err2]]) {
                    expect(err1).to.be.null;
                    expect(err2).to.be.null;
                    next();
                });
            },
            function(next) {
                const tasks = [
                    ['get',key1],
                    ['get',key2],
                    [RedisHelper.EMPTY_FUN],
                ];
                redisHelper.doParallelJobs(tasks,function(none,[[err1,reply1],[err2,reply2]]) {
                    expect(err1).to.be.null;
                    expect(err2).to.be.null;
                    expect(reply1).to.be.equal(value1);
                    expect(reply2).to.be.equal(value2);
                    next();
                });
            }
        ],done);
        
    });

    it('the key should been expired in next day\'s ' + EXPIRE_HOUR + ' hour', function (done) {
        const KEY = 'mykey';
        async.waterfall([
            function(next) {
                redisClient.set(KEY,'some value', function(err) {
                    next(err);
                });
            },
            function(next) {
                redisHelper.expire(KEY,function(err) {
                    next(err);
                });
            },
            function(next) {
                const date = new Date();
                const now = date.getTime();
                date.setDate(date.getDate() + 1);
                date.setHours(EXPIRE_HOUR);
                const expireSecond = Math.floor((date - now) / 1000);
                redisClient.ttl(KEY,function(err, reply) {
                    if (err) {
                        return next(err);
                    }
                    expect(Math.abs(Number(reply) - expireSecond)).to.be.lt(3);
                    next();
                });
            },
            function(next) {
                redisHelper.expire(KEY,function(err, reply) {
                    if (err) {
                        return next(err);
                    }
                    expect(reply).to.be.equal(RedisHelper.HAS_CALL_EXPIRE_BEFORE);
                    next();
                });
            }
        ], done);
        
    });
});



test/mocha/redis_batch_cmd_test.js
const chai = require('chai');
const Redis = require('ioredis');
const expect = chai.expect;
const async = require('neo-async');
const {
    redisBatchCmd: {
        EVENT_ONE_LOOP_FINISHED,
        EVENT_SEND_ERROR,
        BatchHincr,
        BatchZincrby
    }
} = require('../../');
const redisClient = new Redis();//connect to the redis server of localhost:6379

const LOOP_COUNT = 100;
const INTERVAL = 200;
describe('batch command test', function() {
    it('hincr single memeber in single key',function(done) {
        const cmd = new BatchHincr({redisClient,loopInterval:INTERVAL});
        const key = ('test:' + Math.random()).replace('.','');
        cmd.on(EVENT_ONE_LOOP_FINISHED,function() {
            redisClient.hget(key,'name',function(err,reply) {
                if (err) {
                    return done(err);
                }
                expect(Number(reply)).to.be.equal(LOOP_COUNT);
                done();
            });
        });
        cmd.on(EVENT_SEND_ERROR,function(err) {
            done(err);
        });
        
        for(var i=0;i<LOOP_COUNT;i++) {
            cmd.addData(key,1,'name');
        }

    });

    it('hincr multiple memebers in single key',function(done) {
        const cmd = new BatchHincr({redisClient,loopInterval:INTERVAL});
        const key = ('test:' + Math.random()).replace('.','');
        cmd.on(EVENT_ONE_LOOP_FINISHED,function() {
            async.times(LOOP_COUNT,function(index,next) {
                redisClient.hget(key,'name'+index,function(err,reply) {
                    if (err) {
                        return next(err);
                    }
                    expect(Number(reply)).to.be.equal(1);
                    next();
                });
            },done);
        });
        cmd.on(EVENT_SEND_ERROR,function(err) {
            done(err);
        });
       
        for(var i=0;i<LOOP_COUNT;i++) {
            cmd.addData(key,1,'name'+i);
        }
    });

    it('hincr multiple members of multiple keys',function(done) {
        const cmd = new BatchHincr({redisClient,loopInterval:INTERVAL});
        const key = ('test:' + Math.random()).replace('.','');
        const INNER_LOOP_COUNT = 10;
        const SCORE = 3;
        cmd.on(EVENT_ONE_LOOP_FINISHED,function() {
            async.times(LOOP_COUNT,function(index,next) {
                redisClient.hget(key,'name'+index,function(err,reply) {
                    if (err) {
                        return next(err);
                    }
                    expect(Number(reply)).to.be.equal(SCORE*INNER_LOOP_COUNT);
                    next();
                });
            },done);
        });
        cmd.on(EVENT_SEND_ERROR,function(err) {
            done(err);
        });
       
        for(var i=0;i<LOOP_COUNT;i++) {
            for (var j=0;j<INNER_LOOP_COUNT;j++) {
                cmd.addData(key,SCORE,'name'+i);
            }
        }
    });

    it('zincr single memeber in single key',function(done) {
        const cmd = new BatchZincrby({redisClient,loopInterval:INTERVAL});
        const key = ('test:' + Math.random()).replace('.','');
        cmd.on(EVENT_ONE_LOOP_FINISHED,function() {
            redisClient.zscore(key,'name',function(err,reply) {
                if (err) {
                    return done(err);
                }
                expect(Number(reply)).to.be.equal(LOOP_COUNT);
                done();
            });
        });
        cmd.on(EVENT_SEND_ERROR,function(err) {
            done(err);
        });
        
        for(var i=0;i<LOOP_COUNT;i++) {
            cmd.addData(key,1,'name');
        }

    });

    it('zincr multiple members in single key',function(done) {
        const cmd = new BatchZincrby({redisClient,loopInterval:INTERVAL});
        const key = ('test:' + Math.random()).replace('.','');
        cmd.on(EVENT_ONE_LOOP_FINISHED,function() {
            async.times(LOOP_COUNT,function(index,next) {
                redisClient.zscore(key,'name'+index,function(err,reply) {
                    if (err) {
                        return next(err);
                    }
                    expect(Number(reply)).to.be.equal(1);
                    next();
                });
            },done);
        });
        cmd.on(EVENT_SEND_ERROR,function(err) {
            done(err);
        });
       
        for(var i=0;i<LOOP_COUNT;i++) {
            cmd.addData(key,1,'name'+i);
        }
    });

    it('zincr multiple members of multiple keys',function(done) {
        const cmd = new BatchZincrby({redisClient,loopInterval:INTERVAL});
        const key = ('test:' + Math.random()).replace('.','');
        
        const INNER_LOOP_COUNT = 10;
        const SCORE = 3;
        cmd.on(EVENT_ONE_LOOP_FINISHED,function() {
            async.times(LOOP_COUNT,function(index,next) {
                redisClient.zscore(key,'name'+index,function(err,reply) {
                    if (err) {
                        return next(err);
                    }
                    expect(Number(reply)).to.be.equal(SCORE*INNER_LOOP_COUNT);
                    next();
                });
            },done);
        });
        cmd.on(EVENT_SEND_ERROR,function(err) {
            done(err);
        });
       
        for(var i=0;i<LOOP_COUNT;i++) {
            for (var j=0;j<INNER_LOOP_COUNT;j++) {
                cmd.addData(key,SCORE,'name'+i);
            }
        }
    });
});

lib/RedisHelper.js
const async = require('neo-async');
const FlashCache = require('flash-cache');
const EMPTY_FUN = 'empty_fun';
const HAS_CALL_EXPIRE_BEFORE = -9;
/**
 * @typeof ResultArray
 * @type {Array}
 * @property {Error} 0 - the error to return.
 * @property {String|Object} 1 - the value to return.
 */

/**
 * @function ParallelArrayCallback
 * 
 * @param {Boolean} allDone it always be true
 * @param {ResultArray[]} results 
 */
/**
 * @function ParallelObjectCallback
 * 
 * @param {Boolean} allDone it always be true
 * @param {Object.<String,ResultArray>} results 
 */
/**
 * redis helper class
 */
class RedisHelper {
    /**
     * 
     * @param {Object} option 
     * @param {Object} option.redisClient
     * @param {Number} option.expiredHour The hour in next day, it will used in #expire function to expire the given key in next day.
     * @param {Number} option.expiredCacheTime The milliseconds which used by RedisHelper to save the status of calling expire. when it greater than 0, RedisHelper will not call redis' expire command in `option.expiredCacheTime` milliseconds after you call `expire` function.
     */
    constructor({redisClient, expiredHour, expiredCacheTime}) {
        this._redisClient = redisClient;
        this._expiredHour = expiredHour;
        this._expiredCache =expiredCacheTime > 0 ? new FlashCache({expiredCacheTime}) : null;
    }
    /**
     * @private
     * @param {Array[]} task 
     */
    _redisStepFun(task) {
        const redisClient = this._redisClient;
        return function callbackParallel(callback) {
            const [fun,...params] = task;
            if (typeof params[0] === 'boolean') {
                if (redisClient[fun].name !== 'newRedisCommand') {
                    console.error('key type error!');
                }
            }
            if (fun === EMPTY_FUN) {
                return callback(false,params);
            }
            params.push(function parallelCallback(err,reply) {
                callback(false,[err,reply]);
            });
            redisClient[fun].apply(redisClient,params);
        };
    
    }
    /**
     * Do redis requests in parallel, it will continue even if one of request fails.
     *
     * @param {Array|Object} tasks
     * @param {ParallelArrayCallback|ParallelObjectCallback} callback If the tasks is a type of Array, callback will be a ParallelArrayCallback, otherwise it is a ParallelObjectCallback.
     */
    doParallelJobs(tasks,callback) {

        const _this = this;
        if (Array.isArray(tasks)) {
            tasks = tasks.map(function(task) {
                return _this._redisStepFun(task);
            });
        } else {
            const keys = Object.keys(tasks);
            for (var key of keys) {
                tasks[key] = _this._redisStepFun(tasks[key]);
            }
        }
    
        async.parallel(tasks,callback);
    }
    /**
     * Set the key's age, it will be expired in the next day's certain hour which given in the constructor function.
     * 
     * @param {String} key The key you wanna expire
     * @param {Function} callback The callback function of redis
     */
    expire(key, callback) {
        const date = new Date();
        const now = date.getTime();
        date.setDate(date.getDate() + 1);
        date.setHours(this._expiredHour);
        const expireTime = Math.floor((date - now) / 1000);
        const expiredCache = this._expiredCache;
        if (expiredCache) {
            const ele = expiredCache.getElement(key, true);
            if (ele && ele.value) {
                return callback(null, HAS_CALL_EXPIRE_BEFORE);
            }
        }
        this._redisClient.expire(key, expireTime, function(err, reply) {
            if (err) {
                return callback(err);
            }
            if (expiredCache) {
                expiredCache.save(key, expireTime);
            }
            callback(null, reply);
        });
    }
}
/**
 * The empty function to help you to insert empty operation among normal operation.
 * 
 * @constant
 * @type {String}
 */
RedisHelper.EMPTY_FUN = EMPTY_FUN;
/**
 * The flag to indicate whether you have call expire command on given key.
 * 
 * @constant
 * @type {Number}
 */
RedisHelper.HAS_CALL_EXPIRE_BEFORE = HAS_CALL_EXPIRE_BEFORE;
module.exports = RedisHelper;



lib/redis_batch_cmd.js
const EventEmitter = require('events');
/**
 * @namespace redisBatchCmd
 */
/**
 * @constant {String} EVENT_ONE_LOOP_FINISHED The event triggered when send a batch of commands finish.
 * @memberof redisBatchCmd
 */
const EVENT_ONE_LOOP_FINISHED = exports.EVENT_ONE_LOOP_FINISHED = 'one_loop_finished';
/**
 * @constant {String} EVENT_SEND_ERROR The event tirggered when send command to redis fail.
 * @memberof redisBatchCmd
 */
const EVENT_SEND_ERROR = exports.EVENT_SEND_ERROR = 'send_error';
/**
 * @constant {String} EVENT_EXPIRE_ERROR The event triggered when set expire time fail.
 * @memberof redisBatchCmd
 */
const EVENT_EXPIRE_ERROR = exports.EVENT_EXPIRE_ERROR = 'expired_error';

/**
 * @typedef BatchOption
 * @memberof redisBatchCmd
 * 
 * @param {Object} redisClient The instance of ioredis client.
 * @param {String} cmd The name of redis command.
 * @param {Number} [loopInterval=200] The milliseconds of the interval to send a batch of commands.
 * @param {Number} [expireTime] The seconds used to set the redis key.
 */

/**
 * The class for sending batch redis commands.
 * @class
 * 
 * @memberof redisBatchCmd
 */
const RedisBatchCmd = exports.RedisBatchCmd = class RedisBatchCmd extends EventEmitter {
    /**
     * The constructor of RedisBatchCmd
     * @param {BatchOption} option 
     */
    constructor({
        redisClient,cmd,loopInterval=200,expireTime=7200
    }={}) {
        super();
        if (!cmd) {
            throw new Error('cmd should be given');
        }
        this._redisClient = redisClient;
        this._valueMap = new Map();
        this._loopInterval = loopInterval;
        this._cmd = cmd;
        
        this._expireTime = expireTime;
        this._doLoop();
    }
    /**
     * Add data to interval map.
     * @abstract
     * 
     * @param {String} key 
     * @param {Number} score 
     * @param {String} value 
     */
    addData(key,score,value) {
        throw new Error('not supported');
    }
    /**
     * @private
     * @fires RedisBatchCmd#EVENT_SEND_ERROR
     * @fires RedisBatchCmd#EVENT_ONE_LOOP_FINISHED
     * @fires RedisBatchCmd#EVENT_EXPIRE_ERROR
     */
    _doSendPipeline() {
        const redisClient = this._redisClient;
        var keyLen = this._valueMap.size;
        const _this = this;

        const expireTime = this._expireTime;
        for (const [key,data] of this._valueMap) {
            redisClient.pipeline(data).exec(function finish(err) {
                if (err) {
                    /**
                     * @event RedisBatchCmd#EVENT_SEND_ERROR
                     * @type {Error} err
                     * @type {String} key
                     * @type {String|Object} data
                     */
                    _this.emit(EVENT_SEND_ERROR,err,key,data);
                }
                keyLen--;
                if (keyLen === 0) {
                    /**
                     * @event RedisBatchCmd#EVENT_ONE_LOOP_FINISHED
                     */
                    _this.emit(EVENT_ONE_LOOP_FINISHED);
                }
            });
            redisClient.expire(key, expireTime, function(err) {
                /**
                 * @event RedisBatchCmd#EVENT_EXPIRE_ERROR
                 * @type {Error} err
                 */
                _this.emit(EVENT_EXPIRE_ERROR,err);
            });
        }
        this._valueMap.clear();
    }
    /**
     * @private
     */
    _doLoop() {
        const _this = this;
        setTimeout(function batchLoop() {
            _this._doSendPipeline();
            _this._doLoop();
        },this._loopInterval);
    }

};
/**
 * The utility class to send zincr command in batch.
 * @class
 * @memberof redisBatchCmd
 */
exports.BatchZincrby = class BatchZincrby extends RedisBatchCmd {
    constructor(option) {
        option.cmd = 'zincrby';
        super(option);
    }
    addData(key,score,value) {
        const map = this._valueMap;
        const data = (map.get(key) || []) ;
        data.push(['zincrby',key,score,value]);
        map.set(key,data);
    }
    
};
/**
 * The utility class to send hincr command in batch.
 * @class
 * @memberof redisBatchCmd
 */
exports.BatchHincr = class BatchHincr extends RedisBatchCmd {
    constructor(option) {
        option.cmd = 'hincrby';
        super(option);
    }
    addData(key,score,value) {
        const map = this._valueMap;
        const data = (map.get(key) || []) ;
        data.push(['hincrby',key,value,score]);
        map.set(key,data);
    }

};


lib/redis_batch_incr.js
const EventEmitter = require('events');
/**
 * @namespace RedisBatchIncr
 */
/**
 * @constant {String} EVENT_ONE_LOOP_FINISHED The event triggered when send a batch of commands finish.
 * @memberof RedisBatchIncr
 */
const EVENT_ONE_LOOP_FINISHED = exports.EVENT_ONE_LOOP_FINISHED = 'one_loop_finished';
/**
 * @constant {String} EVENT_SEND_ERROR The event tirggered when send command to redis fail.
 * @memberof RedisBatchIncr
 */
const EVENT_SEND_ERROR = exports.EVENT_SEND_ERROR = 'send_error';
/**
 * @constant {String} EVENT_EXPIRE_ERROR The event triggered when set expire time fail.
 * @memberof RedisBatchIncr
 */
const EVENT_EXPIRE_ERROR = exports.EVENT_EXPIRE_ERROR = 'expired_error';

/**
 * @typedef BatchOption
 * @memberof RedisBatchIncr
 * 
 * @param {Object} redisClient The instance of ioredis client.
 * @param {String} cmd The name of redis command.
 * @param {Number} [loopInterval=200] The milliseconds of the interval to send a batch of commands.
 * @param {Number} [expireTime] The seconds used to set the redis key.
 */

/**
 * The class for sending batch redis commands.
 * @class
 * 
 * @memberof RedisBatchIncr
 */
const RedisBatchIncr = exports.RedisBatchIncr = class RedisBatchIncr extends EventEmitter {
    /**
     * The constructor of RedisBatchIncr
     * @param {BatchOption} option 
     */
    constructor({
        redisClient,cmd,loopInterval=200,expireTime=7200,key
    }={}) {
        super();
        if (!cmd) {
            throw new Error('cmd should be given');
        }
        this._redisClient = redisClient;
        this._valueMap = new Map();
        this._loopInterval = loopInterval;
        this._cmd = cmd;
        this._key = key;
        
        this._expireTime = expireTime;
        this._doLoop();
    }
    /**
     * Add data to interval map.
     * 
     * 
     * @param {String} key 
     * @param {Number} score 
     * @param {String} value 
     */
    addData(score,value) {

        const oldScore = this._valueMap.get(value) || 0;
        this._valueMap.set(value, oldScore + score);
    }
    _buildCmdParams() {
        throw new Error('not implemented');
    }
    /**
     * @private
     * @fires RedisBatchIncr#EVENT_SEND_ERROR
     * @fires RedisBatchIncr#EVENT_ONE_LOOP_FINISHED
     * @fires RedisBatchIncr#EVENT_EXPIRE_ERROR
     */
    _doSendPipeline() {
        const redisClient = this._redisClient;
        var valueLen = this._valueMap.size;
        const _this = this;

        const expireTime = this._expireTime;
        var i = 0;
        const data = new Array(valueLen);
        for (const [value, score] of this._valueMap) {
            data[i] = this._buildCmdParams (value, score);
            i++;
        }
        redisClient.pipeline(data).exec(function finish(err) {
            if (err) {
                /**
                 * @event RedisBatchIncr#EVENT_SEND_ERROR
                 * @type {Error} err
                 * @type {String|Object} data
                 */
                _this.emit(EVENT_SEND_ERROR,err,data);
            }

            /**
             * @event RedisBatchIncr#EVENT_ONE_LOOP_FINISHED
             * @type {Number} len The length of members
             */
            _this.emit(EVENT_ONE_LOOP_FINISHED, valueLen);
        });
            
        
        redisClient.expire(this._key, expireTime, function(err) {
            /**
             * @event RedisBatchIncr#EVENT_EXPIRE_ERROR
             * @type {Error} err
             */
            _this.emit(EVENT_EXPIRE_ERROR,err);
        });
        this._valueMap.clear();
    }
    /**
     * @private
     */
    _doLoop() {
        const _this = this;
        setTimeout(function batchLoop() {
            _this._doSendPipeline();
            _this._doLoop();
        },this._loopInterval);
    }

};
/**
 * The utility class to send zincr command in batch.
 * @class
 * @extends RedisBatchIncr
 */
exports.BatchZincrby = class BatchZincrby extends RedisBatchIncr {
    constructor(option) {
        option.cmd = 'zincrby';
        super(option);
    }

    _buildCmdParams(value, score) {
        return [this._cmd , this._key, score, value];
    }
};
/**
 * The utility class to send hincr command in batch.
 * @class
 * @extends RedisBatchIncr
 */
exports.BatchHincr = class BatchHincr extends RedisBatchIncr {
    constructor(option) {
        option.cmd = 'hincrby';
        super(option);
    }

    _buildCmdParams( value, score) {
        return [this._cmd , this._key, value, score];
    }
};


lib/FlashCacheRedis.js
const DEFAULT_CACHE_AGE = 1000;
const FlashCache = require('flash-cache');
/**
 * @class FlashCachRedis
 * 
 * A tool that cache the data in memory to reduce the frequency of reading redis. It use the package of [flash-cache](https://www.npmjs.com/package/flash-cache) inside.
 * 
 * @property {Object} cacheQueryAdapter An wrapper of redis query fucntion , such as `get` `hget` `smembers` `zrange`.
 * @example
 * ```javascript
 * const {FlashCacheRedis} = require('redis-speed');
 * const Redis = require('ioredis');
 * const redisClient = new Redis();//connect to the redis server of localhost:6379
 * const {cacheQueryAdapter} = new FlashCacheRedis({
 *    redisClient,//the redis client object
 *    interval:1000
 * });
 * const KEY = 'flash_cache_redis:basic';
 * const VALUE = {name:'sunny',id:1};
 * 
 * //First query the `key` in memory, if not have, query in redis. 
 * cacheQueryAdapter.get(key,function(err,reply) {});
 * ```
 */
class FlashCachRedis {
    /**
     * 
     * @param {Object} option
     * @param {Redis} option.redisClient
     * @param {Number} [option.interval=1000] The milliseconds of lifecyle of the data saved in memory
     * @param {Boolean} [option.useYoungOnly] Whether only reading the data from young area , it will passed to the instance of `FlashCache`.
     */
    constructor({redisClient, interval = DEFAULT_CACHE_AGE, useYoungOnly = false}) {
        this.redisClient = redisClient;
        this.cacheResult = new FlashCache({interval});
        this._useYoungOnly = useYoungOnly;

        const commands = redisClient.getBuiltinCommands();
        const _this = this;
        this.cacheQueryAdapter = {};
        for (let command of commands) {
            this.cacheQueryAdapter[command] = function newRedisCommand(...args) {
                _this.queryWithCache(command,...args);
            };
        }
    }
    queryWithCache(fun,...args) {
        let saveFlag = true;
        if (typeof args[0] === 'boolean') {
            saveFlag = args.shift();
        }
        const len = args.length;
        const lastParam = args[len -1];
        let callback = function(){};
        let paramCount = 0;
        if (typeof(lastParam) === 'function') {
            callback = lastParam;
            paramCount = len -1;
        } else {
            paramCount = len;
        }
        let itemName = fun;
        const queryParams = new Array(paramCount+1);
        for (let i=0;i<paramCount;i++) {
            const p = args[i];
            itemName += ':' + p;
            queryParams[i] = p;
        }
        const cacheResult = this.cacheResult;
        if (saveFlag) {
            const element = cacheResult.getElement(itemName, this._useYoungOnly);
            if (element) {
                return callback(null,element.value);
            }
        }
        queryParams[paramCount] = function redisCallback(err,reply) {
            callback(err,reply);
            if (saveFlag) {
                cacheResult.save(itemName,reply);
            }
        };

        this.redisClient[fun](...queryParams);
    }
}

module.exports = FlashCachRedis;